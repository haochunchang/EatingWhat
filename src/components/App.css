/* #### Generated By: http://www.cufonfonts.com #### */

@font-face {
  font-family: "ObelixPro Regular";
  font-style: normal;
  font-weight: normal;
  src: local("ObelixPro Regular"),
    url("../font/FontsFreeNetObelixProcyr.woff") format("woff");
  font-display: swap;
}

h1 {
  font-family: "ObelixPro Regular";
  font-weight: 500;
}

.App {
  text-align: center;
}

.center {
  display: flex;
  justify-content: center;
}

section {
  padding: 1em;
}
