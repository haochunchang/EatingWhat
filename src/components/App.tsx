import React, { useEffect, useState } from "react";
import { TypedUseSelectorHook, useSelector, useDispatch } from "react-redux";

import "./App.css";

import { requestFoods } from "../actions";
import { FOOD_GITHUB_URL } from "../api/api";

import Scroll from "./Scroll";
import ErrorBoundary from "../containers/ErrorBoundary";
import FoodButton from "./FoodButton";
import CardList from "./CardList";

export interface Food {
  id: number;
  name: string;
  category: string;
  url: string;
}

export interface AppState {
  requestFoods: {
    foods: Array<Food>;
    isPending: boolean;
    error: string;
  };
}

export function filterFoods(foods: Array<Food>) {
  return {
    starchyFoods: foods.filter((item) => item.category === "main food"),
    proteinFoods: foods.filter((item) => item.category === "meat"),
  };
}

export default function App(): JSX.Element {
  const typedUseSelector: TypedUseSelectorHook<AppState> = useSelector;
  const isPending = typedUseSelector((state) => state.requestFoods.isPending);
  const foods = typedUseSelector((state) => state.requestFoods.foods);

  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(requestFoods(FOOD_GITHUB_URL));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const { starchyFoods, proteinFoods } = filterFoods(foods);
  const [starch, setStarch] = useState(starchyFoods);
  const [protein, setProtein] = useState(proteinFoods);

  useEffect(() => {
    setStarch(starchyFoods);
    setProtein(proteinFoods);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [foods]);

  const onClickChangeStarch = (): void => {
    const index = Math.floor(Math.random() * starchyFoods.length);
    setStarch([starchyFoods[index]]);
  };

  const onClickChangeProtein = (): void => {
    const index = Math.floor(Math.random() * proteinFoods.length);
    setProtein([proteinFoods[index]]);
  };

  return isPending ? (
    <div className="App">
      <header className="header">
        <h1 className="f1">Loading...</h1>
      </header>
    </div>
  ) : (
    <div className="App">
      <header className="header">
        <h1 className="f1">Eating What?</h1>
      </header>
      <div className="row">
        <div className="column">
          <FoodButton
            id="starch-button"
            onClicked={onClickChangeStarch}
            FoodName={"Starchy Food"}
          />
          <Scroll>
            <ErrorBoundary>
              <CardList foods={starch} />
            </ErrorBoundary>
          </Scroll>
        </div>

        <div className="column">
          <FoodButton
            id="protein-button"
            onClicked={onClickChangeProtein}
            FoodName={"Protein Food"}
          />
          <Scroll>
            <ErrorBoundary>
              <CardList foods={protein} />
            </ErrorBoundary>
          </Scroll>
        </div>
      </div>
    </div>
  );
}
