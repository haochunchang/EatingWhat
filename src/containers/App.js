import React, { Component } from "react";
import logo from "../logo.svg";
import "./App.css";
import Button from "../components/Button";
import CardList from "../components/CardList";
import SearchBox from "../components/SearchBox";
import Scroll from "../components/Scroll";

class App extends Component {
  constructor() {
    super();
    this.state = {
      foods: [],
      searchfield: "",
      index: -1,
    };
  }

  componentDidMount() {
    fetch(
      "https://my-json-server.typicode.com/haochunchang/food-json-server/posts/"
    )
      .then((response) => response.json())
      .then((posts) => this.setState({ foods: posts }));
  }

  onSearchChange = (event) => {
    this.setState({ searchfield: event.target.value });
  };

  onClicked = (event) => {
    const random = Math.floor(Math.random() * this.state.foods.length + 1);
    this.setState({ index: random });
    console.log(event.target.value);
  };

  render() {
    const { foods, searchfield, index } = this.state;
    var filteredFoods = [];
    if (index < 0) {
      filteredFoods = foods.filter((food) => {
        return food.name.toLowerCase().includes(searchfield.toLowerCase());
      });
    } else {
      filteredFoods = foods.filter((food) => food.id === index);
    }
    return !foods.length ? (
      <div className="App">
        <header className="header">
          <h1 className="f1">Loading...</h1>
        </header>
      </div>
    ) : (
      <div className="App">
        <header className="header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="f1">What Do You Want To Eat?</h1>
        </header>
        <div className="tc">
          <SearchBox searchChange={this.onSearchChange} />
          <Button />
          <Scroll>
            <CardList foods={filteredFoods} />
          </Scroll>
        </div>
      </div>
    );
  }
}

export default App;
