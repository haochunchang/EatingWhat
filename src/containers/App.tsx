import React, { Component } from "react";
import MainPage from "../components/MainPage";
import { AppState, AppProps, Food } from "../components/MainPage";

import { connect } from "react-redux";
import { setSearchField, requestFoods } from "../actions";
import { FOOD_URL } from "../constants";

interface Dispatch {
  (dispatch: any): Promise<any>;
  type?: string;
  payload?: any;
}

const mapStateToProps = (state: AppState) => {
  return {
    searchfield: state.searchFoods.searchfield,
    starchyFoods: state.requestFoods.foods.filter(
      (item) => (item as Food).category === "main food"
    ),
    proteinFoods: state.requestFoods.foods.filter(
      (item) => (item as Food).category !== "main food"
    ),
    isPending: state.requestFoods.isPending,
    error: state.requestFoods.error,
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    onSearchChange: (event: React.SyntheticEvent<HTMLInputElement>) =>
      dispatch(setSearchField((event.target as HTMLInputElement).value)),
    onRequestFoods: () => dispatch(requestFoods(FOOD_URL)),
  };
};

class App extends Component<AppProps, AppState> {
  render() {
    return <MainPage {...this.props} />;
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
