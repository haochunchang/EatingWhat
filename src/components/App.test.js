import React from "react";
import { Provider } from "react-redux";
import { shallow, mount, configure } from "enzyme";
import { act } from "react-dom/test-utils";
import Adapter from "enzyme-adapter-react-16";

import configureMockStore from "redux-mock-store";
import thunkMiddleware from "redux-thunk";

import App, { filterFoods } from "./App";

configure({ adapter: new Adapter() });

const mockStore = configureMockStore([thunkMiddleware]);
const store = mockStore({
  requestFoods: {
    foods: [
      {
        name: "Ice cream",
        category: "main food",
        url: "",
      },
      {
        name: "Pork",
        category: "meat",
        url: "",
      },
    ],
    isPending: false,
    error: "",
  },
});

const pendingStore = mockStore({
  requestFoods: {
    foods: [],
    isPending: true,
    error: "",
  },
});

describe("App testing", () => {
  let shallowWrapper;
  let wrapper;

  beforeAll(() => {
    shallowWrapper = shallow(
      <Provider store={store}>
        <App />
      </Provider>
    );
    wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });

  it("renders App correctly", () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("renders Pending App correctly", () => {
    const pendingWrapper = shallow(
      <Provider store={pendingStore}>
        <App />
      </Provider>
    );
    expect(pendingWrapper).toMatchSnapshot();
  });

  it("should update starch foods on click", () => {
    const starchButton = wrapper.find("#starch-button").hostNodes();
    const handleClick = jest.spyOn(starchButton.props(), "onClick");
    act(() => starchButton.prop("onClick")());
    expect(handleClick).toHaveBeenCalled();
  });

  it("should update protein foods on click", () => {
    const proteinButton = wrapper.find("#protein-button").hostNodes();
    const handleClick = jest.spyOn(proteinButton.props(), "onClick");
    act(() => proteinButton.prop("onClick")());
    expect(handleClick).toHaveBeenCalled();
  });

  it("tests filterFood", () => {
    const foods = [
      { name: "rice", category: "main food" },
      { name: "pork", category: "meat" },
    ];
    const filtered = filterFoods(foods);
    expect(filtered).toHaveProperty("starchyFoods", [foods[0]]);
    expect(filtered).toHaveProperty("proteinFoods", [foods[1]]);
  });
});
