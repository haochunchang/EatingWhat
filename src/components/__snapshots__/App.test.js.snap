// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App testing renders App correctly 1`] = `ShallowWrapper {}`;

exports[`App testing renders Pending App correctly 1`] = `ShallowWrapper {}`;
